### PR TITLE
fix: ignore keys from client persistence which already have commit ids

### LINKED ISSUE #731: Empty batch requests being sent to server
**Describe the bug**
When uncommitted entries contains the keys are not synced to secondary server (the cached keys and keys that are local to client), 
the sync service skips the entries. But this causes the delay keys to sync because the non sync keys in uncommitted entries list is being iterated and an empty batch requests are being sent to server.

Client Logs:

```
  FINER|2022-09-29 07:20:25.629525|SyncService|cached:@ce2e2:phonebeef3ec1-7696-4f11-ac7a-76851125cca6.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629552|SyncService|cached:@ce2e2:phone55df8976-f61f-4fb3-b547-56bbcfae9aaf.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629567|SyncService|cached:@ce2e2:phone094f55ce-cd83-4218-bb6a-43e38cc53471.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629576|SyncService|cached:@ce2e2:phone540142bc-a9ce-4c7a-aa91-e3d3babd7304.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629589|SyncService|cached:@ce2e2:phonea20590dd-c87c-43a3-8e65-253c4b4cc66b.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629636|AtLookup|SENDING: batch:[]
   
  
  FINER|2022-09-29 07:20:25.708053|OutboundMessageListener|RECEIVED data:[] 
  
  FINER|2022-09-29 07:20:25.717212|SyncService|batch result:data:[] 
```

Server Logs:

```
ce2e2_secondary.1.vtcdxu0eh32m@cicd2x64    | INFO|2022-09-29 07:20:26.287185|InboundListener|RCVD: [_431fb7f9-5fe5-4fa6-9d67-b0e3ebc1574b] batch:[] 
ce2e2_secondary.1.vtcdxu0eh32m@cicd2x64    | 
ce2e2_secondary.1.vtcdxu0eh32m@cicd2x64    | INFO|2022-09-29 07:20:26.287396|InboundConnectionImpl|SENT: [_431fb7f9-5fe5-4fa6-9d67-b0e3ebc1574b] data:[]
```

<details>
<summary> Complete logs describing the delay in sync process</summary>

```
FINER|2022-09-29 07:20:25.629180|SyncService|syncing to remote. Total uncommitted entries: 423 
  
  FINER|2022-09-29 07:20:25.629525|SyncService|cached:@ce2e2:phonebeef3ec1-7696-4f11-ac7a-76851125cca6.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629552|SyncService|cached:@ce2e2:phone55df8976-f61f-4fb3-b547-56bbcfae9aaf.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629567|SyncService|cached:@ce2e2:phone094f55ce-cd83-4218-bb6a-43e38cc53471.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629576|SyncService|cached:@ce2e2:phone540142bc-a9ce-4c7a-aa91-e3d3babd7304.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629589|SyncService|cached:@ce2e2:phonea20590dd-c87c-43a3-8e65-253c4b4cc66b.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629636|AtLookup|SENDING: batch:[]
   
  
  FINER|2022-09-29 07:20:25.708053|OutboundMessageListener|RECEIVED data:[] 
  
  FINER|2022-09-29 07:20:25.717212|SyncService|batch result:data:[] 
  
  FINER|2022-09-29 07:20:25.717265|SyncService|cached:@ce2e2:phoned3643ecb-897b-427e-b973-a61e307a5c34.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.717284|SyncService|cached:@ce2e2:phonea04e879a-ceb8-4dbf-a0ad-70612ba37a7b.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
    
  FINER|2022-09-29 07:20:28.357460|SyncService|batchId:1 key:_latestnotificationidv2.keystream@ce2e2 
  
  FINER|2022-09-29 07:20:28.357674|SyncService|batchId:2 key:public:publickey@ce2e2 
  
  FINER|2022-09-29 07:20:28.357849|SyncService|batchId:3 key:_latestnotificationidv2.wavi@ce2e2 
  
  FINER|2022-09-29 07:20:28.358006|SyncService|batchId:4 key:_latestnotificationidv2.wavi@ce2e2 
  
  FINER|2022-09-29 07:20:28.358163|SyncService|batchId:5 key:_latestnotificationidv2.wavi@ce2e2 
  
  FINER|2022-09-29 07:20:28.358238|AtLookup|SENDING: batch:[{"id":1,"command":"update:ttl:0:ttb:0:ccd:false:isBinary:false:isEncrypted:false:_latestnotificationidv2.keystream@ce2e2 mv0p9rEK2+rgQ8RG1/qiofXjqxFyLZItQUicQWvWjqvmSEe1Pl35QqWgdSwlLyd97KXnLJrvhd00vVuJNvmWlu0br8wcmBvnYerJdd5RHSXiZKMn1jUTLSVyoRHB+qsSs35kcYKJV00OAlAZQXiyusJQ5XgV1hC+t7ZKSYyfiIDNo423y64LqZUpwh1TZ+tI0qA0JBfwKKvXSsMTPMVsViHyG3YDE7pi8j7LRs4PSVroqilJumYEZnfLHw/epMcpOov7r877HO+wBER/3C9312xyd/XpjSYTM+7yP8AJoeODYkgga5dbnI9KcSIXlQs1v2+jPRUVahQ9W2DQOwmpUkWAe9o0q272xBqNDwcVc1qC3NcMhjLTxdB7wtP+ENiiXG615IuNfCh/kQ+87wrzTQ=="***{"id":2,"command":"update:ttl:0:ttb:0:ccd:false:isBinary:false:isEncrypted:false:public:publickey@ce2e2 MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAls/anFf/q9OdIIGICEwEosKHVgDe8GiVMLWmdAbVVYdORAcUsWedNYJ5fuWBtiToaQE/NXB4e8J3CgpbQVeoOgwdKM7DTl7mS0JT10bgxVaKf01EYE9+4I2/sVpRKe/rvQ3kr2y5k5mPIxm/0GkMRidYRTriEu1bEky152RLEjLNXTLBnI/kvWJDApkbD9FwMyBFZBRHjb2kg/jV+eDpv+IFl+uLO/tdpIc+hpa/PsSP6uL9KSJKL03p0BqYQik3oAbdEdN2eyU/mYanmVeRC/YDnGo4OtGjtBni9N4qHavQKE9aMMU2gB5K0JuxRZwdmH8U8TdsR3oeDjMAHT3oGwIDAQAB"***{"id":3,"command":"update:isBinary:false:isEncrypted:false:_latestnotificationidv2.wavi@ce2e2 mv0p9rEK2+q1FpMQhqryofDhphdyLcQkR0jEQTmCjqDuTRW2P1qkHv75JSwlLyd97KXnLJrvhd00vVuONamfkLgcpMwWn0m3YerLdd5RHSflYqNx3jhCLiJxrRLLo/EU6CxhetWIUk1dUFhEQXiyv5ZQ5XxF1hDv5LVPGIrD0YbM89C3y64LqZUpwh1TZ+tI0qA0JBfwKKvXSsMTPMVsViHyG3YDE7pi8j7LRs4PSVroqilJumYEZnfLHw/epMcpOov7r877HOq1D0l+2CF312xyd/XpjSYTM+7yP8AJoeODYkgga5dbnI9KcSIXlQs1v2+jPRUVahQ9W2DQOwmpUkWAe9o0q272xBqNDwcVc1qC3NcMhjLTxdB7wtP+ENiiXG615IuNfCh/kQ+87wrzTQ=="***{"id":4,"command":"update:isBinary:false:isEncrypted:false:_latestnotificationidv2.wavi@ce2e2 mv0p9rEK2+q1FpMQhqryofDhphdyLcQkR0jEQTmCjqDuTRW2P1qkHv75JSwlLyd97KXnLJrvhd00vVuONamfkLgcpMwWn0m3YerLdd5RHSflYqNx3jhCLiJxrRLLo/EU6CxhetWIUk1dUFhEQXiyv5ZQ5XxF1hDv5LVPGIrD0YbM89C3y64LqZUpwh1TZ+tI0qA0JBfwKKvXSsMTPMVsViHyG3YDE7pi8j7LRs4PSVroqilJumYEZnfLHw/epMcpOov7r877HOq1D0l+2CF312xyd/XpjSYTM+7yP8AJoeODYkgga5dbnI9KcSIXlQs1v2+jPRUVahQ9W2DQOwmpUkWAe9o0q272xBqNDwcVc1qC3NcMhjLTxdB7wtP+ENiiXG615IuNfCh/kQ+87wrzTQ=="***{"id":5,"command":"update:isBinary:false:isEncrypted:false:_latestnotificationidv2.wavi@ce2e2 mv0p9rEK2+q1FpMQhqryofDhphdyLcQkR0jEQTmCjqDuTRW2P1qkHv75JSwlLyd97KXnLJrvhd00vVuONamfkLgcpMwWn0m3YerLdd5RHSflYqNx3jhCLiJxrRLLo/EU6CxhetWIUk1dUFhEQXiyv5ZQ5XxF1hDv5LVPGIrD0YbM89C3y64LqZUpwh1TZ+tI0qA0JBfwKKvXSsMTPMVsViHyG3YDE7pi8j7LRs4PSVroqilJumYEZnfLHw/epMcpOov7r877HOq1D0l+2CF312xyd/XpjSYTM+7yP8AJoeODYkgga5dbnI9KcSIXlQs1v2+jPRUVahQ9W2DQOwmpUkWAe9o0q272xBqNDwcVc1qC3NcMhjLTxdB7wtP+ENiiXG615IuNfCh/kQ+87wrzTQ=="***]
   
  
  FINEST|2022-09-29 07:20:28.367926|Monitor|_messageHandler received data 
  
  FINER|2022-09-29 07:20:28.367973|Monitor|received response on monitor: notification: {"id":"43cddf1e-bb9a-4c56-a20c-209a268b858a","from":"@ce2e1","to":"@ce2e2","key":"@ce2e2:phonenumber02d85ecf-ba87-4945-860e-844ab7b9a93a.wavi@ce2e1","value":null,"operation":"delete","epochMillis":1664436028280,"messageType":"MessageType.key","isEncrypted":false*** 
  

```
</details>

Github action run : https://github.com/atsign-foundation/at_client_sdk/actions/runs/3149341230/jobs/5120863575

### PR DESCRIPTION
**Pull request by srieteja**
closes https://github.com/atsign-foundation/at_client_sdk/issues/731

**- What I did**
- At the moment, getChanges() in commitLogKeyStore returns all the uncommitted entries from the given sequence number. However, there is a bug in getChanges() which on the client side returns entries that are already synced (entries that already have a commitID) due to which the entries are again sent back to the server. Further, if the entries returned contain cached keys, empty batch requests are sent (since cached keys are not synced local -> remote).
- The fix would be when fetching the un-committed entries have a check to ignore the entries that have commitId, since the entries are already in sync with the server.

**-Impact of this fix**
- This fix ensures on the client-side that already committed entries are not returned by getChanges() thus reducing the number of redundant entries being returned. This would resolve the case where empty batches are being synced. An added benefit would be the reduced redundant entries that are being input into other processes like sync; hence relatively improving efficiency.
- This fix does not affect the functioning of getChanges() on the server side.

**- How I did it**
- added check in class CommitLogKeystore and method getChanges() that ensures only  keys without commit ids are returned when the enableCommitId flag is false

**- How to verify it**
- tests added in commit_log_test.dart to assert the changes
- when the getChanges() method is called on the client persistence, keys with already existing commit ids should not be returned.

**- Description for the changelog**

fix: ignore keys from client persistence which already have commit ids

### CODE REVIEW

File: packages/at_persistence_secondary_server/lib/src/log/commitlog/commit_log_keystore.dart [hunk 1]
```diff
@@ -315,15 +315,41 @@ class CommitLogKeyStore
           if (element.key >= startKey &&
               _acceptKey(element.atKey, regexString) &&
               changes.length <= limit) {
-            changes.add(element);
+            switch (enableCommitId) {
```
Comment (sitaram-kalluri) [inline packages/at_persistence_secondary_server/lib/src/log/commitlog/commit_log_keystore.dart]: How about replacing line no. 313 to 352 as below:

The idea behind this refactoring is to remove the redundant code inside the for loop with and without the limit populated.

When the limit is null, set it to a max int value.

```dart
      // If limit is null, assign it to max int number.
      limit ??= double.maxFinite as int;
      for (CommitEntry element in values) {
        if (element.key >= startKey &&
            _acceptKey(element.atKey!, regexString) &&
            changes.length <= limit) {
          // On the client side, add the entries whose commit Id's are null
          // Checking the enableCommitId is false to ensure the commit log instance is of client
          if (enableCommitId == false) {
            changes.any((element) => element.commitId == null);
          } else {
            changes.add(element);
          }
        }
      }
      return changes;
```

**Review by gkc — CHANGES_REQUESTED** [review]
Can you explain in detail please in the PR summary *why* these changes will fix the linked bug?

Please also include your analysis of the impact these changes would have on other code flows, and why these changes are 'safe'

**Comment by murali-shris** [comment]
@srieteja please write tests for the changes you have done.

File: packages/at_persistence_secondary_server/lib/src/log/commitlog/commit_log_keystore.dart (hunk 1 above)
Comment (srieteja) [reply packages/at_persistence_secondary_server/lib/src/log/commitlog/commit_log_keystore.dart]: this snipped has been included in the code as part of https://github.com/atsign-foundation/at_server/pull/1020/commits/0d4248ef5abbe50ca1f9f76ae30a08df354fb527

**Comment by srieteja** [comment]
> @srieteja please write tests for the changes you have done.

tests have been added as part of https://github.com/atsign-foundation/at_server/pull/1020/commits/adbcd31f6401cfde514fbf12506d1fc49c641eb2. Also tested this on my local and the empty-batch sync does not seem to be happening.

**Review by gkc — APPROVED** [review]
LGTM

**Review by sitaram-kalluri — APPROVED** [review]
